Infinite loop for doctesting: caching issue with atomic

I'm not exatly sure what's going on here with this docs example:


```
>>> from astropy import units as u
>>> from astroquery.atomic import AtomicLineList
>>> wavelength_range = (15 * u.nm, 1.5e+16 * u.Hz)
>>> AtomicLineList.query_object(wavelength_range=wavelength_range, wavelength_type='Air',
...                             wavelength_accuracy=20, element_spectrum='C II-IV')
```

Running it in ipython I get the expected result (pending some not yet merged changes that get rid of a few warnings):
```
<Table length=3>
LAMBDA VAC ANG SPECTRUM  TT  CONFIGURATION   TERM   J J      A_ki    LEVEL ENERGY  CM 1
   float64       str4   str2     str14       str6   str5   float64         str18       
-------------- -------- ---- -------------- ------ ----- ----------- ------------------
      196.8874     C IV   E1 1s2.2s-1s2.12p 2S-2Po 1/2-*  91300000.0 0.00 -   507904.40
      197.7992     C IV   E1 1s2.2s-1s2.11p 2S-2Po 1/2-* 118000000.0 0.00 -   505563.30
      199.0122     C IV   E1 1s2.2s-1s2.10p 2S-2Po 1/2-* 157000000.0 0.00 -   502481.80
```

Running it with doctest OTOH:
```
========================================================================= FAILURES =========================================================================
___________________________________________________________________ [doctest] atomic.rst ___________________________________________________________________
023 In the following Python session you can see the ``atomic`` package in
024 action. Note that Hz is actually not a supported unit by Atomic Line List,
025 the atomic package takes care to support all spectral units.
026 
027 .. doctest-remote-data::
028 
029     >>> from astropy import units as u
030     >>> from astroquery.atomic import AtomicLineList
031     >>> wavelength_range = (15 * u.nm, 1.5e+16 * u.Hz)
032     >>> AtomicLineList.query_object(wavelength_range=wavelength_range, wavelength_type='Air',
UNEXPECTED EXCEPTION: RecursionError('maximum recursion depth exceeded while calling a Python object')
Traceback (most recent call last):
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/doctest.py", line 1336, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest atomic.rst[3]>", line 1, in <module>
  File "/private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astroquery-test-r53oieer/lib/python3.9/site-packages/astroquery/atomic/core.py", line 151, in query_object
    response = self.query_object_async(
  File "/private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astroquery-test-r53oieer/lib/python3.9/site-packages/astroquery/atomic/core.py", line 246, in query_object_async
    response = self._submit_form(input_payload, cache=cache)
  File "/private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astroquery-test-r53oieer/lib/python3.9/site-packages/astroquery/atomic/core.py", line 299, in _submit_form
    response = self._request("POST", url=url, data=payload,
  File "/private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astroquery-test-r53oieer/lib/python3.9/site-packages/astroquery/query.py", line 324, in _request
    to_cache(response, query.request_file(self.cache_location))
  File "/private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astroquery-test-r53oieer/lib/python3.9/site-packages/astroquery/query.py", line 30, in to_cache
    response = copy.deepcopy(response)
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
...
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 205, in _deepcopy_list
    append(deepcopy(a, memo))
  File "/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/copy.py", line 137, in deepcopy
    d = id(x)
RecursionError: maximum recursion depth exceeded while calling a Python object
/private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astroquery-test-r53oieer/lib/python3.9/site-packages/docs/atomic/atomic.rst:32: UnexpectedException
================================================================= short test summary info ==================================================================
FAILED docs/atomic/atomic.rst::atomic.rst
=============================================================== 1 failed, 9 passed in 3.35s ================================================================
```

All the while if I add the `cache=False` kwarg the doctesting passes as expected, too.

I haven't seen this type of issues with any of the other modules, so it may not be a generic caching issue, but something specific to atomic.

--- a/astroquery/atomic/core.py
+++ b/astroquery/atomic/core.py
@@ -220,7 +220,7 @@ class AtomicLineListClass(BaseQuery):
         if upper_level_erange is not None:
             upper_level_erange = upper_level_erange.to(
                 u.cm ** -1, equivalencies=u.spectral()).value
-        input = {
+        input_payload = {
             'wavl': ' '.join(map(str, wlrange_in_angstroms)),
             'wave': 'Angstrom',
             'air': air,
@@ -242,8 +242,8 @@ class AtomicLineListClass(BaseQuery):
             'tptype': 'as_a',
         }
         if get_query_payload:
-            return input
-        response = self._submit_form(input, cache=cache)
+            return input_payload
+        response = self._submit_form(input_payload, cache=cache)
         return response
 
     def _parse_result(self, response):
@@ -259,7 +259,7 @@ class AtomicLineListClass(BaseQuery):
         colnames = [colname.strip('-').replace('-', ' ')
                     for colname in header.split('|') if colname.strip()]
         indices = [i for i, c in enumerate(header) if c == '|']
-        input = []
+        result_data = []
         for line in data:
             row = []
             for start, end in zip([0] + indices, indices + [None]):
@@ -272,26 +272,26 @@ class AtomicLineListClass(BaseQuery):
                     # maintain table dimensions when data missing
                     row.append('None')
             if row:
-                input.append('\t'.join(row))
-        if input:
-            return ascii.read(input, data_start=0, delimiter='\t',
+                result_data.append('\t'.join(row))
+        if result_data:
+            return ascii.read(result_data, data_start=0, delimiter='\t',
                               names=colnames, fast_reader=False)
         else:
             # return an empty table if the query yielded no results
             return Table()
 
-    def _submit_form(self, input=None, cache=True):
+    def _submit_form(self, input_data=None, cache=True):
         """Fill out the form of the SkyView site and submit it with the
-        values given in `input` (a dictionary where the keys are the form
+        values given in ``input_data`` (a dictionary where the keys are the form
         element's names and the values are their respective values).
 
         """
-        if input is None:
-            input = {}
-        # only overwrite payload's values if the `input` value is not None
+        if input_data is None:
+            input_data = {}
+        # only overwrite payload's values if the ``input_data`` value is not None
         # to avoid overwriting of the form's default values
-        payload = self._default_form_values.copy()
-        for k, v in input.items():
+        payload = self.__default_form_values.copy()
+        for k, v in input_data.items():
             if v is not None:
                 payload[k] = v
         url = self._form_action_url

--- a/astroquery/atomic/core.py
+++ b/astroquery/atomic/core.py
@@ -302,8 +302,8 @@ class AtomicLineListClass(BaseQuery):
         return response
 
     def _default_form_values(self):
-        response = self._request("GET", url=self.FORM_URL, data={},
-                                 timeout=self.TIMEOUT, cache=True)
+        response = self._request("GET", url=self.FORM_URL, params={},
+                                 timeout=self.TIMEOUT, cache=False)
         bs = BeautifulSoup(response.text, features='html5lib')
         form = bs.find('form')
         default_form_values = self._get_default_form_values(form)

--- a/astroquery/atomic/core.py
+++ b/astroquery/atomic/core.py
@@ -294,24 +294,22 @@ class AtomicLineListClass(BaseQuery):
         for k, v in input.items():
             if v is not None:
                 payload[k] = v
-        url = urlparse.urljoin(self.FORM_URL, self._default_form.get('action'))
+        url = self._form_action_url
         log.debug(f"final payload = {payload} from url={url}")
         response = self._request("POST", url=url, data=payload,
                                  timeout=self.TIMEOUT, cache=cache)
         log.debug("Retrieved data from POST request")
         return response
 
-    @property
     def _default_form_values(self):
+        response = self._request("GET", url=self.FORM_URL, data={},
+                                 timeout=self.TIMEOUT, cache=True)
+        bs = BeautifulSoup(response.text, features='html5lib')
+        form = bs.find('form')
+        default_form_values = self._get_default_form_values(form)
+        self._form_action_url = urlparse.urljoin(self.FORM_URL, form.get('action'))
 
-        if self.__default_form_values is None:
-            response = self._request("GET", url=self.FORM_URL, data={},
-                                     timeout=self.TIMEOUT, cache=True)
-            bs = BeautifulSoup(response.text, features='html5lib')
-            self._default_form = form = bs.find('form')
-            self.__default_form_values = self._get_default_form_values(form)
-
-        return self.__default_form_values
+        return default_form_values
 
     def _get_default_form_values(self, form):
         """Return the already selected values of a given form (a BeautifulSoup

--- a/astroquery/atomic/core.py
+++ b/astroquery/atomic/core.py
@@ -247,7 +247,7 @@ class AtomicLineListClass(BaseQuery):
         return response
 
     def _parse_result(self, response):
-        data = StringIO(BeautifulSoup(response.text, 'html5').find('pre').text.strip())
+        data = StringIO(BeautifulSoup(response.text, features='html5lib').find('pre').text.strip())
         # `header` is e.g.:
         # "u'-LAMBDA-VAC-ANG-|-SPECTRUM--|TT|--------TERM---------|---J-J---|----LEVEL-ENERGY--CM-1----'"
         # `colnames` is then
@@ -307,7 +307,7 @@ class AtomicLineListClass(BaseQuery):
         if self.__default_form_values is None:
             response = self._request("GET", url=self.FORM_URL, data={},
                                      timeout=self.TIMEOUT, cache=True)
-            bs = BeautifulSoup(response.text, 'html5')
+            bs = BeautifulSoup(response.text, features='html5lib')
             self._default_form = form = bs.find('form')
             self.__default_form_values = self._get_default_form_values(form)
 

--- a/astroquery/atomic/core.py
+++ b/astroquery/atomic/core.py
@@ -27,7 +27,7 @@ class AtomicLineListClass(BaseQuery):
 
     def __init__(self):
         super(AtomicLineListClass, self).__init__()
-        self.__default_form_values = self._default_form_values()
+        self.__default_form_values = None
 
     def query_object(self, *, wavelength_range=None, wavelength_type=None, wavelength_accuracy=None, element_spectrum=None,
                      minimal_abundance=None, depl_factor=None, lower_level_energy_range=None,
@@ -179,7 +179,7 @@ class AtomicLineListClass(BaseQuery):
         response : `requests.Response`
             The HTTP response returned from the service.
         """
-        default_values = self.__default_form_values
+        default_values = self._default_form_values()
         wltype = (wavelength_type or default_values.get('air', '')).lower()
         if wltype in ('air', 'vacuum'):
             air = wltype.capitalize()
@@ -290,7 +290,7 @@ class AtomicLineListClass(BaseQuery):
             input_data = {}
         # only overwrite payload's values if the ``input_data`` value is not None
         # to avoid overwriting of the form's default values
-        payload = self.__default_form_values.copy()
+        payload = self._default_form_values().copy()
         for k, v in input_data.items():
             if v is not None:
                 payload[k] = v
@@ -302,14 +302,17 @@ class AtomicLineListClass(BaseQuery):
         return response
 
     def _default_form_values(self):
-        response = self._request("GET", url=self.FORM_URL, params={},
-                                 timeout=self.TIMEOUT, cache=False)
-        bs = BeautifulSoup(response.text, features='html5lib')
-        form = bs.find('form')
-        default_form_values = self._get_default_form_values(form)
-        self._form_action_url = urlparse.urljoin(self.FORM_URL, form.get('action'))
-
-        return default_form_values
+        if self.__default_form_values is None:
+            response = self._request("GET", url=self.FORM_URL, params={},
+                                     timeout=self.TIMEOUT, cache=False)
+            bs = BeautifulSoup(response.text, features='html5lib')
+            form = bs.find('form')
+            default_form_values = self._get_default_form_values(form)
+            self._form_action_url = urlparse.urljoin(self.FORM_URL, form.get('action'))
+            self.__default_form_values = default_form_values
+        else:
+            raise ValueError(self.__default_form_values)
+        return self.__default_form_values
 
     def _get_default_form_values(self, form):
         """Return the already selected values of a given form (a BeautifulSoup

--- a/astroquery/atomic/core.py
+++ b/astroquery/atomic/core.py
@@ -179,7 +179,7 @@ class AtomicLineListClass(BaseQuery):
         response : `requests.Response`
             The HTTP response returned from the service.
         """
-        default_values = self._default_form_values
+        default_values = self.__default_form_values
         wltype = (wavelength_type or default_values.get('air', '')).lower()
         if wltype in ('air', 'vacuum'):
             air = wltype.capitalize()
@@ -197,8 +197,8 @@ class AtomicLineListClass(BaseQuery):
             raise ValueError('Invalid parameter "transitions": {0!r}'
                              .format(transitions))
         if transitions is None:
-            _type = self._default_form_values.get('type')
-            type2 = self._default_form_values.get('type2')
+            _type = default_values.get('type')
+            type2 = default_values.get('type2')
         else:
             s = str(transitions)
             if len(s.split(',')) > 1:

--- a/astroquery/atomic/core.py
+++ b/astroquery/atomic/core.py
@@ -29,7 +29,7 @@ class AtomicLineListClass(BaseQuery):
         super(AtomicLineListClass, self).__init__()
         self.__default_form_values = self._default_form_values()
 
-    def query_object(self, wavelength_range=None, wavelength_type=None, wavelength_accuracy=None, element_spectrum=None,
+    def query_object(self, *, wavelength_range=None, wavelength_type=None, wavelength_accuracy=None, element_spectrum=None,
                      minimal_abundance=None, depl_factor=None, lower_level_energy_range=None,
                      upper_level_energy_range=None, nmax=None, multiplet=None, transitions=None,
                      show_fine_structure=None, show_auto_ionizing_transitions=None,
@@ -167,7 +167,7 @@ class AtomicLineListClass(BaseQuery):
         return table
 
     @prepend_docstr_nosections(query_object.__doc__)
-    def query_object_async(self, wavelength_range=None, wavelength_type='', wavelength_accuracy=None,
+    def query_object_async(self, *, wavelength_range=None, wavelength_type='', wavelength_accuracy=None,
                            element_spectrum=None, minimal_abundance=None, depl_factor=None,
                            lower_level_energy_range=None, upper_level_energy_range=None, nmax=None, multiplet=None,
                            transitions=None, show_fine_structure=None, show_auto_ionizing_transitions=None,

--- a/astroquery/atomic/core.py
+++ b/astroquery/atomic/core.py
@@ -27,7 +27,7 @@ class AtomicLineListClass(BaseQuery):
 
     def __init__(self):
         super(AtomicLineListClass, self).__init__()
-        self.__default_form_values = None
+        self.__default_form_values = self._default_form_values()
 
     def query_object(self, wavelength_range=None, wavelength_type=None, wavelength_accuracy=None, element_spectrum=None,
                      minimal_abundance=None, depl_factor=None, lower_level_energy_range=None,

--- a/astroquery/atomic/core.py
+++ b/astroquery/atomic/core.py
@@ -161,7 +161,8 @@ class AtomicLineListClass(BaseQuery):
             output_columns=output_columns, cache=cache,
             get_query_payload=get_query_payload)
         if get_query_payload:
-            return input
+            return response
+
         table = self._parse_result(response)
         return table
 

